### PR TITLE
`Programming exercises`: Generate solution entries for behavioral test cases

### PR DESCRIPTION
**Pull request by Hialus**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] ~~I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).~~ Testwise coverage not supported here

### Motivation and Context

These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
A previous PR already introduced the ability to create solution entries for structural test cases. This PR tackles solution entries for behavioral test cases using the git-diff and testwise coverage reports.
A reminder: Solution entries are a single code block that the students have to change in the template to get closer to passing a test case.

### Description

The process of generating solution entries for behavioral test cases is a bit complicated, but I'll try to give a brief rundown below. Due to this and the fact that the process should be easily extendable in the future, I used a slightly modified Blackboard pattern for this. This means that all steps are split up into knowledge sources and new knowledge sources can easily be added in the future.

Requirements:
- The git-diff report and testwise coverage report must exist
- The solution build plan must have run through without a build failure

Steps:
1. `GroupGitDiffAndCoverageEntriesByFilePathAndTestCase`: This will create GroupedFiles that bundle all ProgrammingExerciseGitDiffEntries and TestwiseCoverageReportEntries of a single file and test case together
2. `ExtractCoveredLines`: A small step that extracts all lines that were covered by the TestwiseCoverageReportEntries
3. `ExtractChangedLines`: A small step that extracts all lines that were changed in the ProgrammingExerciseGitDiffEntries
4. `FindCommonLines`: The intersection of the previous two lists
5. `CreateCommonChangeBlocks`: Takes the output of the previous step and creates GroupedFile.ChangeBlock instances from it. Each code block represents a single string of continuous common lines
6. `InsertFileContents`: Inserts the contents of each file in the solution repository into their related GroupedFiles
7. `AddUncoveredLinesAsPotentialCodeBlocks`: There are certain lines that are not covered by Jacoco, but are still relevant for the student or can improve the readability of the generated solution entries. These lines include else statements, curly brackets, and empty lines. 
This step will add those lines as potential code blocks, which will only be added in certain circumstances (This will be part of a future PR. Currently all code blocks are included in the end result)
8. `CombineChangeBlocks`: The previous step can lead to directly adjacent or intersecting code blocks within a single grouped file. This step combines those code blocks into one.
9. `CreateSolutionEntries`: Creates a solution entry for every code block in every grouped file

### Steps for Testing

As there still is not proper view for solution entries in the client you will again need to rely on the developer console output. #4976 Will include the first part of this, so if that PR gets merged you can use that.

Prerequisites:
- 1 Admin

1. Log in to Artemis
2. Create a Gradle_Gradle Java Programming exercise with Testwise Coverage option enabled
3. Do any changes to the test and solution repo that you want
4. Wait for the build of the solution build plan to finish
5. Open the developer console
6. Press the "Create behavioral solution entries button"
7. Check that the printed solution entries are correct (Cross-reference with the git-diff and coverage reports and under consideration of the AddUncoveredLinesAsPotentialCodeBlocks step)

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| All Knowledge sources | >80% | >77% |
| BehavioralBlackboard | n/a | 100% |
| BehavioralSolutionEntryGenerationException | n/a | 50% |
| BehavioralTestCaseService | 73% | 88% |
| GroupedFile | 58% | 98% |
| GroupedFile.ChangeBlock | 61% | 72% |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java
Comment (ole-ve) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java:68]: Why does the check for a potential prefix start two lines before the first line?

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java
Comment (ole-ve) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java:85]: Aside from the for-loop condition, all lines are equal for `getPotentialPrefix` and `getPotentialPostfix`.

A possibility to resolve this could be to define a private interface with a method that returns the line number (index) if the condition is still valid and -1 if the condition is not fulfilled anymore? -1 can never be a valid value for line numbers and therefore could be used for this.
The functionality could be extracted into a new method that additionally has this interface as a parameter. In `getPotentialPrefix` and `getPotentialPostfix` you could then hand over a lamba function describing this modified condition behavior.
What do you think about that idea?

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/CombineChangeBlocks.java
Comment (ole-ve) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/CombineChangeBlocks.java:37]: Condition for checking if this is the last block, right? I'd appreciate a one-liner comment

**Review by ole-ve — CHANGES_REQUESTED** [review]
Great work, I really like the application of the Blackboard Pattern! 👍
Breaks down the generation well into smaller parts and makes it more maintainable for the future.

I reviewed the code and added some inline comments.
I also tested the changes manually on my local setup and on TS3 with the template exercise and a real exercise, the entries were created correctly. There is room for improvement, but thanks to the implementation of the Blackboard Pattern, this can be covered in follow-up PR(s).

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/FindCommonLines.java
Comment (bassner) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/FindCommonLines.java:33]: Just stumbled over this. ``not all match`` is not the same as ``none match``. Not sure if you intended to change the actual behavior of this line or just the used method. Do you actually want this (which isn't any shorter so I'd leave the original)?
```java
return blackboard.getGroupedFiles().stream().anyMatch(groupedFile -> !groupedFile.getCommonLines().isEmpty());
```

Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/FindCommonLines.java:33]: I agree with Patrick. This change would break the method.

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/CreateCommonChangeBlocks.java
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/CreateCommonChangeBlocks.java:41]: You are correct. I will add a comment.

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/CreateCommonChangeBlocks.java
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/CreateCommonChangeBlocks.java:50]: Yes it is. I will add a comment

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java:68]: Added a comment

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java:85]: This would make the code quite a lot more complicated IMO. So I would leave it as it is instead of introducing a new interface just for this.

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/CombineChangeBlocks.java
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/CombineChangeBlocks.java:37]: Added a comment

File: src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/behavioral/knowledgesource/AddUncoveredLinesAsPotentialCodeBlocks.java:86]: Fine by me 👍

**Review by ole-ve — APPROVED** [review]
Approve for code and re-approve for manual testing, because only comments were added since my last review 👍

**Review by ge65cer — DISMISSED** [review]
Testing Session: Duplicate entries, as old structural solution entries are not deleted

**Review by willergeorg — DISMISSED** [review]
Tested on TS3. No errors occurred and reasonable output was displayed in the dev console.

**Review by ole-ve — APPROVED** [review]
Tested locally, the entries get deleted as expected.
Code still looks good to me 👍

**Review by Dominik-Weinzierl — APPROVED** [review]
Code looks good to me. All newly added endpoints have authorities and the tests look also good. Well done!

**Review by ole-ve — APPROVED** [review]
Re-approve manual testing & code review after merge

**Review by willergeorg — APPROVED** [review]
Only merge conflict resolved since last review.

**Review by Dominik-Weinzierl — APPROVED** [review]
Re-Approve

**Pull request merged**
